app/src/main/java/at/tacticaldevc/oat/utils/Prefs.java

```
@@ -27,16 +34,125 @@
+        ensureNotNull(context, "Application Context");
+        ensureStringIsValid(password, "new User Password");
+
+        SharedPreferences prefs = context.getSharedPreferences(DOCUMENT_NAME_DATA, Context.MODE_PRIVATE);
+        SharedPreferences.Editor editor = prefs.edit();
+
+        byte[] salt;
+        MessageDigest algorithm;
+        try {
+            algorithm = MessageDigest.getInstance("SHA-256");
+        } catch (NoSuchAlgorithmException ex) {
+            throw OATApplicationException.forLibraryError("MessageDigest", ex);
+        }
+
+        if (prefs.getString(KEY_COMMAND_PASSWORD_SALT, null) == null) {
+            SecureRandom rnd = new SecureRandom();
+            salt = new byte[16];
+            rnd.nextBytes(salt);
+            editor.putString(KEY_COMMAND_PASSWORD_SALT, Base64.encodeToString(salt, Base64.NO_WRAP));
+        } else {
+            salt = Base64.decode(prefs.getString(KEY_COMMAND_PASSWORD_SALT, null), Base64.NO_WRAP);
+        }
+        algorithm.update(salt);
+        String hash = Base64.encodeToString(algorithm.digest(password.getBytes()), Base64.NO_WRAP);
+
+        editor.putString(KEY_COMMAND_PASSWORD, hash);
+        editor.apply();
+    }
+
+    /**
+     * Verifies that a given Password is correct
+     *
+     * @param context         the Context of the Application
+     * @param passwordToCheck the password to be verified
+     * @return true if the password matches the password that was stored
+     */
+    public static boolean verifyApplicationPassword(Context context, String passwordToCheck) {
+        ensureNotNull(context, "Application Context");
+        try {
+            ensureStringIsValid(passwordToCheck, "password to check");
+        } catch (IllegalArgumentException ex) {
+            return false;
+        }
+
+        SharedPreferences prefs = context.getSharedPreferences(DOCUMENT_NAME_DATA, Context.MODE_PRIVATE);
+
+        String salt = prefs.getString(KEY_COMMAND_PASSWORD_SALT, "");
```

IMO we should throw an exception if there is no salt saved. It doesn't make sense if there is a password but no salt. Having no salt saved would mean that the password is lost and the data must be exported/deleted

---

implemented change